src/GafferScene/Instancer.cpp

```diff
@@ -269,8 +493,116 @@ class Instancer::EngineData : public Data
 			return result;
 		}
 
+		typedef std::map< InternedString, boost::unordered_set< IECore::MurmurHash > > PrototypeHashes;
+
+		// In order to compute the number of variations, we compute a unique hash for every context we use
+		// for evaluating prototypes.  So that we can track which sources are responsible for variations,
+		// we return a vector of hash sets, corresponding to the order of m_prototypeContextVariables,
+		// plus an extra entry at the end for the combined result of all variation sources
+		std::unique_ptr<PrototypeHashes> uniquePrototypeHashes() const
+		{
+			std::vector< boost::unordered_set< IECore::MurmurHash > > variableHashAccumulate( m_prototypeContextVariables.size() );
+			boost::unordered_set< IECore::MurmurHash > totalHashAccumulate;
+
+			size_t n = numPoints();
+			for( unsigned int i = 0; i < n; i++ )
+			{
+				int protoIndex = prototypeIndex( i );
+				if( protoIndex == -1 )
+				{
+					continue;
+				}
+
+				IECore::MurmurHash totalHash;
+				InternedStringVectorData &rootPath = *m_roots[ protoIndex ];
```

Comment:
Would be nice to use `const &` here to make it clear that we're not mutating anything. Actually, could you redeclare `m_roots` as `std::vector<ConstInternedStringVectorDataPtr>` while you're at it, to force this?

Reply:
Done.